dipy/reconst/mcsd.py

```diff
@@ -0,0 +1,268 @@
+        reg[:iso, :iso] = np.eye(iso)
+        reg[iso:, iso:] = odf_reg
+
+        X = B * multiplier_matrix
+
+        self.fitter = QpFitter(X, reg)
+        self.sh_order = sh_order
+        self._X = X
+        self.sphere = reg_sphere
+        self.response = response
+        self.B_dwi = B
+        self.m = m
+        self.n = n
+
+    def predict(self, params, gtab=None, S0=None):
+        """Compute a signal prediction given spherical harmonic coefficients
+        for the provided GradientTable class instance.
+
+        Parameters
+        ----------
+        params : ndarray
+            The spherical harmonic representation of the FOD from which to make
+            the signal prediction.
+        gtab : GradientTable
+            The gradients for which the signal will be predicted. Use the
+            model's gradient table by default.
+        S0 : ndarray or float
+            The non diffusion-weighted signal value.
+            Default : None
+        """
+        if gtab is None:
+            X = self._X
+        else:
+            iso = self.response.iso
+            B, m, n = multi_tissue_basis(gtab, self.sh_order, iso)
+            multiplier_matrix = _inflate_response(self.response, gtab, n,
+                                                  self.delta)
+            X = B * multiplier_matrix
+        return np.dot(params, X.T)
+
+    @multi_voxel_fit
+    def fit(self, data):
+        coeff = self.fitter(data)
+        return MSDeconvFit(self, coeff, None)
+
+
+class MSDeconvFit(shm.SphHarmFit):
+
+    def __init__(self, model, coeff, mask):
+        """
+        Inherits the SphHarmFit which fits the diffusion data to a spherical
+        harmic model.
+        """
+        self._shm_coef = coeff
+        self.mask = mask
+        self.model = model
+
+    @property
+    def shm_coeff(self):
+        return self._shm_coef[..., self.model.response.iso:]
+
+    @property
+    def volume_fractions(self):
+        tissue_classes = self.model.response.iso + 1
+        return self._shm_coef[..., :tissue_classes] / SH_CONST
+
+
+def quadprog(P, Q, G, H):
+    r"""
+    Helper function to set up and solve the Quadratic Program (QP) in CVXPY.
+    A QP problem has the following form:
+    minimize      1/2 x' P x + Q' x
+    subject to    G x <= H
+
+    Here the QP solver is based on CVXPY and uses OSQP.
+    """
+    x = cvx.Variable(Q.shape[0])
+    P = cvx.Constant(P)
+    objective = cvx.Minimize(0.5 * cvx.quad_form(x, P) + Q * x)
+    constraints = [G*x <= H]
+
+    # setting up the problem
+    prob = cvx.Problem(objective, constraints)
+    prob.solve()
+    opt = np.array(x.value).reshape((Q.shape[0],))
+    return opt
+
+
+class QpFitter(object):
+
+    def __init__(self, X, reg):
+        r"""
+        Makes use of the quadratic programming solver `quadprog` to fit the
+        model. The initialization for the model is done using the warm-start by
+        default in `CVXPY`.
```

Comment:
Would be good to document the parameters.

Reply:
Looks okay?

Reply:
Still no documentation here. What is `X`? What is `reg`? Are they expected to have a particular type/shape?